flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/SchedulingStrategyUtils.java

```diff
@@ -81,6 +82,14 @@
 			final SchedulingTopology topology,
 			final Set<SchedulingPipelinedRegion> regions) {
 
+		// Avoid the O(V) (V is the number of vertices in the topology) sorting
+		// complexity if the given set of regions is small enough
+		if (regions.size() == 0) {
+			return Collections.emptyList();
+		} else if (regions.size() == 1) {
+			return Collections.singletonList(regions.iterator().next());
+		}
+
 		return IterableUtils.toStream(topology.getVertices())
```

Comment:
maybe we should also convert stream to a loop for performance in `sortPipelinedRegionsInTopologicalOrder`?

Reply:
I think we can keep it as is because it is not the critical part. 
The reworked java streams are in per-region code path, while this one is invoked the per-event code path. But even the reworked loops are not most critical, because it is in a O(V) complexity.
The other 2 commits to fix the vertex/region sorting, reduces O(V^2) complexity to O(V) in each valid triggering event.